statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+                // N.B. According to the snap documentation, this compession should never fail for
+                // our use case.
+                let compressed_write_request = match compress_write_request(write_request) {
+                    Ok(c) => c,
+                    Err(e) => {
+                        stats.requests_compression_fail.inc();
+                        error!(
+                            "failed to compress prometheus WriteRequest, dropping samples: {:?}",
+                            e
+                        );
+                        return;
+                    }
+                };
+                stats
+                    .bytes_sent_total
+                    .inc_by(compressed_write_request.len() as f64);
+
+                // Attempt each request at least once
+                let mut res = client
+                    .send_write_request(compressed_write_request.clone())
+                    .await;
+                // And retry only if the tripwire has not yet been triggered.
+                while should_retry(&res) {
+                    sleep(Duration::from_millis(retry_delay_ms)).await;
+                    stats.requests_retry.inc();
+                    tokio::select! {
+                        new_res = client.send_write_request(compressed_write_request.clone()) => {
+                            res = new_res;
+                        },
+                        _ = trip.clone() => {
+                            break;
+                        },
+                    }
+                }
+
+                if let Err(e) = res {
+                    stats.requests_fail.inc();
+                    stats.samples_send_fail.inc_by(samples_in_batch);
+                    warn!(
+                        "prometheus remote write request failed, dropping samples: {:?}",
+                        e
+                    );
+                }
+
+                drop(permit);
+                drop(flush_tx);
+            });
+        }
+
+        if shutdown {
+            // Wait for all spawned tasks to terminate.
+            drop(flush_tx);
+            let _ = flush_rx.recv().await;
+            return;
+        }
+    }
+}
+
+async fn construct_batch(
+    rx: &mut mpsc::Receiver<Pdu>,
+    timeout: Duration,
+    max_batch_size: usize,
+) -> (Vec<Pdu>, bool) {
+    let mut batch = Vec::<Pdu>::with_capacity(max_batch_size);
+    let sleep = sleep(timeout);
+    tokio::pin!(sleep);
+    loop {
+        select! {
+            opdu = rx.recv() => {
+                if let Some(pdu) = opdu {
+                    batch.push(pdu);
+                    if batch.len() >= max_batch_size {
+                        return (batch, false);
+                    }
+                } else {
+                    return (batch, true)
+                }
+            }
+            _ = &mut sleep => return (batch, false)
+        }
+    }
+}
+
+fn construct_write_request(stats: &Stats, batch: Vec<Pdu>) -> WriteRequest {
```

Comment:
This isn't something you need to do anything about, just something to keep in mind: this function allocates a bit.

My guess is it's completely fine, but if you find yourself needing to reduce allocator pressure or decrease peak heap, I'd look in here since there's a lot of `strcpy` behind the owned string construction.